bin/manage.sh

```diff
@@ -0,0 +1,236 @@
+    return $?
+}
+
+doHealthCheck() {
+    local status=$(couchbase-cli server-info ${CB_CONN} | jq -r .status)
+    if [[ $status != "healthy" ]]; then
+       echo "Node not healthy, status was: $status"
+       return 1
+    fi
+    return 0
+}
+
+
+# -------------------------------------------
+# Joining a cluster
+
+
+# We only need one IP from the healthy cluster in order to join it.
+getHealthyClusterIp() {
+    echo $(curl -Lsf http://${CONSUL}:8500/v1/health/service/${COUCHBASE_SERVICE_NAME}?passing | jq -r .[0].Service.Address)
+}
+
+# If we fail to join the cluster, then bail out and hit it on the
+# next health check
+joinCluster(){
+    echo '# Joining cluster...'
+    local node=$1
+    curl -Lsif -u ${COUCHBASE_USER}:${COUCHBASE_PASS} \
+         -d "hostname=${IP_PRIVATE}&user=admin&password=password" \
+         "http://${node}:8091/controller/addNode" || exit 1
+    echo 'Joined cluster!'
+    rebalance
+    exit 0
+}
+
+# We need to rebalance for each node because we can't guarantee that we won't
+# try to rebalance while another node is coming up. Doing this in a loop because
+# we can't queue-up rebalances -- the rebalance command cannot be called while a
+# rebalance is in progress
+rebalance() {
+    echo '# Rebalancing cluster...'
+    while true; do
+        echo -n '.'
+        couchbase-cli rebalance ${CB_CONN} && return
+        sleep .7
+    done
+}
+
+
+# -------------------------------------------
+# Bootstrapping a cluster
+
+# Try to obtain a lock in Consul. If we can't get the lock then another node
+# is trying to bootstrap the cluster. The cluster-init node will have 120s
+# to show up as healthy in Consul.
+obtainBootstrapLock() {
+    echo 'No cluster nodes found, trying to obtain lock on bootstrap...'
+    local session=$(curl -Lsf -XPUT -d '{"Name": "couchbase-bootstrap", "TTL": "120s"}' http://${CONSUL}:8500/v1/session/create | jq -r .ID) || return $?
+    local lock=$(curl -Lsf -XPUT http://${CONSUL}:8500/v1/kv/couchbase-bootstrap?acquire=$session)
+    if [[ $lock == "true" ]]; then
+        return 0
+    else
+        return 1
+    fi
+}
+
+# bootstrap the Couchbase cluster and set resource limits
+initCluster() {
+    echo
+    echo '# Bootstrapping cluster...'
+
+    # Couchbase resource limits
+    local avail_memory=$(free -m | grep -o "Mem:\s*[0-9]*" | grep -o "[0-9]*")
+    local cb_memory=$((($avail_memory/10)*7))
+    local avail_cpus=$(nproc)
+    local cb_cpus=$(($avail_cpus>8?8:$avail_cpus))
+    local cb_cpus=$(($cb_cpus>1?$cb_cpus:1))
+
+    couchbase-cli cluster-init -c 127.0.0.1:8091 -u access -p password \
+                  --cluster-init-username=${COUCHBASE_USER} \
+                  --cluster-init-password=${COUCHBASE_PASS} \
+                  --cluster-init-port=8091 \
+                  --cluster-init-ramsize=${cb_memory} \
+                  --services=data,index,query
+
+    couchbase-cli bucket-create ${CB_CONN} \
```

Comment:
The problem is that we don't have the Couchbase server running during `preStart`, and all of the Couchbase CLI tools talk to a Couchbase server rather than manipulating the on-disk configuration. We did a lot of this in `preStart` in the MySQL autopilot pattern and it involved starting up and tearing down the server before starting it again. That's going to be particularly painful to do cleanly with a multi-process server like Couchbase.